clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java

```diff
@@ -235,4 +373,105 @@ public void ack(final long currentTimeMs) {
+            Map<TopicPartition, OffsetAndMetadata> offsets = new HashMap<>(responseData.size());
+            Set<TopicPartition> unstableTxnOffsetTopicPartitions = new HashSet<>();
+            for (Map.Entry<TopicPartition, OffsetFetchResponse.PartitionData> entry : responseData.entrySet()) {
+                TopicPartition tp = entry.getKey();
+                OffsetFetchResponse.PartitionData partitionData = entry.getValue();
+                if (partitionData.hasError()) {
+                    Errors error = partitionData.error;
+                    log.debug("Failed to fetch offset for partition {}: {}", tp, error.message());
+
+                    if (error == Errors.UNKNOWN_TOPIC_OR_PARTITION) {
+                        request.future.completeExceptionally(new KafkaException("Topic or Partition " + tp + " does not " +
+                                "exist"));
+                        return;
+                    } else if (error == Errors.TOPIC_AUTHORIZATION_FAILED) {
+                        if (unauthorizedTopics == null) {
+                            unauthorizedTopics = new HashSet<>();
+                        }
+                        unauthorizedTopics.add(tp.topic());
+                    } else if (error == Errors.UNSTABLE_OFFSET_COMMIT) {
+                        System.out.println("asdkljlsadjflksajfdlk");
+                        unstableTxnOffsetTopicPartitions.add(tp);
+                    } else {
+                        request.future.completeExceptionally(new KafkaException("Unexpected error in fetch offset " +
+                                "response for partition " + tp + ": " + error.message()));
+                        return;
+                    }
+                } else if (partitionData.offset >= 0) {
+                    // record the position with the offset (-1 indicates no committed offset to fetch);
+                    // if there's no committed offset, record as null
+                    offsets.put(tp, new OffsetAndMetadata(partitionData.offset, partitionData.leaderEpoch, partitionData.metadata));
+                } else {
+                    log.info("Found no committed offset for partition {}", tp);
+                    offsets.put(tp, null);
+                }
+            }
+
+            if (unauthorizedTopics != null) {
+                request.future.completeExceptionally(new TopicAuthorizationException(unauthorizedTopics));
+            } else if (!unstableTxnOffsetTopicPartitions.isEmpty()) {
+                System.out.println("oshdofjakl");
+                // just retry
+                log.info("The following partitions still have unstable offsets " +
+                        "which are not cleared on the broker side: {}" +
+                        ", this could be either " +
+                        "transactional offsets waiting for completion, or " +
+                        "normal offsets waiting for replication after appending to local log", unstableTxnOffsetTopicPartitions);
```

Comment:
Unlike `COORDINATOR_LOAD_IN_PROGRESS` which is a request-level global error code, `UNSTABLE_OFFSET_COMMIT` is a partition-level error. Do we always retry all the partitions as long as one of them has this error today?

Reply:
Yeah, that's the current behavior. I'll leave a note here for this optimization idea.

Reply:
Sounds good! Thanks for confirming.